core/commands/pin.go

```diff
@@ -200,25 +212,78 @@ Example:
 			res.SetError(err, cmds.ErrNormal)
 			return
 		}
-		if !found {
-			typeStr = "recursive"
-		}
 
-		switch typeStr {
-		case "all", "direct", "indirect", "recursive":
-		default:
-			err = fmt.Errorf("Invalid type '%s', must be one of {direct, indirect, recursive, all}", typeStr)
-			res.SetError(err, cmds.ErrClient)
+		if found {
+			switch typeStr {
+			case "all", "direct", "indirect", "recursive":
+			default:
+				err = fmt.Errorf("Invalid type '%s', must be one of {direct, indirect, recursive, all}", typeStr)
+				res.SetError(err, cmds.ErrClient)
+			}
 		}
 
 		keys := make(map[string]RefKeyObject)
-		if typeStr == "direct" || typeStr == "all" {
-			for _, k := range n.Pinning.DirectKeys() {
+
+		if len(req.Arguments()) > 0 {
+			if !found {
+				typeStr = "all"
+			}
```

Comment:
since this command is now more involved, let's put the contents of this `if` in a different function (and maybe the other case (len(args)==0) too), to decompose out the logic and make it easier to understand + change in the future:

```go

if len(req.Arguments()) > 0 {
  if !found {
    typeStr = "all"
  }

  keys, err := pinLsKeys(req.Arguments(), typeStr, ...) 
  if err != nil {
    res.SetError(err, cmds.ErrNormal)
  } else {
    res.SetOutput(&RefKeyList{Keys: keys})
  }
  return
}

```